File: Backend-AWSLambda/api_contract.py
import s3_access

ADMIN = "Admin"
PEER = "Peer"
AQUAINTED = "Aquainted"
DISTANT = "Distant"

PNG = "png"
JPG = "jpg"
MP4 = "mp4"
MOV = "mov"

def make_APILocation(node, to_host_weight=DISTANT):
    longitude = node["location"]["default_location"][0]
    latitude = node["location"]["default_location"][1]
    authorized = (to_host_weight == PEER or to_host_weight == ADMIN)
    if authorized and node["location"]["live_location"] is not None:
        longitude = node["location"]["live_location"][0]
        latitude = node["location"]["live_location"][1]
    return [float(longitude), float(latitude)]

def make_APIGeoPoint(api_location):
    return {
        'type': "Feature",
        'properties': None,
        'geometry': {
            'type': "Point",
            'coordinates': api_location
        }
    }

def make_APIGeoTrigger(trigger_id, zoom, api_location):
    return {
        'type': "Feature",
        'properties': {
            'trigger_id': trigger_id,
            'zoom': float(zoom)
        },
        'geometry': {
            'type': "Point",
            'coordinates': api_location
        }
    }

def make_APIGeoNode(node, from_host_weight=DISTANT, to_host_weight=DISTANT):
    live_location_enabled = None
    if from_host_weight == ADMIN and to_host_weight == ADMIN:
        live_location_enabled = node["location"]["live_location"] is not None
    portrait_id = s3_access.presigned_url(node["media"]["portrait_id"])
    supplement_id = s3_access.presigned_url(node["media"]["supplement_id"])
    counter_weight = "Distant" if to_host_weight == DISTANT else "Close"

    return {
        'type': "Feature",
        'properties': {
            'id': node["id"],
            'name': node["name"],
            'description': node["description"],
            'live_location_enabled': live_location_enabled,
            'zoom': float(node["zoom"]),
            'media': {
                'portrait_id': portrait_id,
                'supplement_id': supplement_id
            },
            'weight': from_host_weight,
            'counter_weight': counter_weight
        },
        'geometry': {
            'type': "Point",
            'coordinates': make_APILocation(node, to_host_weight)
        }
    }

def make_APIGeoJSON(api_geos):
    return {
        "type": "FeatureCollection",
        "crs": {
            "type": "name",
            "properties": {
                "name": "urn:ogc:def:crs:OGC:1.3:CRS84"
            }
        },
        "features": api_geos
    }

def make_APICanvasDrop(drop, api_geo_node=None):
    x = drop['canvas_location'][0]
    y = drop['canvas_location'][1]
    z = drop['canvas_location'][2]
    
    image_id = None
    if 'image_id' in drop: 
        image_id = s3_access.presigned_url(drop["image_id"])

    return {
        'id': drop["id"],
        'canvas_location': [float(x), float(y), float(z)],
        'image_id': image_id,
        'portal_url': drop["portal_url"] if 'portal_url' in drop else None,
        'api_geo_node': api_geo_node
    }

def make_APIServerMessage(message):
    return {
        'message': message
    }

def make_APIData(api_obj):
    return {
        'data': api_obj
    }    

File: Frontend-IOS/Common/Network/api_contract.swift
//
//  APIContract.swift
//  Owl (iOS)
//
//  Created by Developer on 7/3/21.
//

import Foundation
import SwiftUI

let API_DOMAIN: URL = URL(string: "https://4s9pe4gqvc.execute-api.us-east-2.amazonaws.com/dev/private")!
let PUBLIC_API_DOMIAN: URL = URL(string: "https://4s9pe4gqvc.execute-api.us-east-2.amazonaws.com/dev/public")!


//MARK: Read
struct APIData<T: Codable>: Codable {
    let data: T
}

struct APIGeoJSON: Codable {
    let type: String
    let crs: CRS
    let features: [APIGeoNode] 
    
    struct CRS: Codable {
        let type: String
        let properties: Properties
        
        struct Properties: Codable {
            let name: String
        }
    }
}

struct APIGeoTrigger: Codable {
    let type: String
    let properties: Properties
    let geometry: APIGeoNode.Geometry
    
    struct Properties: Codable {
        let trigger_id: String
        let zoom: Double
    }
}

struct APIGeoNode: Codable {
    let type: String
    let properties: Properties
    let geometry: Geometry

    struct Properties: Codable {
        let id: String
        let name: String
        let description: String
        let live_location_enabled: Bool?
        let zoom: Double
        let media: Media
        let weight: Weight
        let counter_weight: CounterWeight
        
        struct Media: Codable {
            let portrait_id: String
            let supplement_id: String
        }
        
        enum Weight: String, Codable {
            case Admin
            case Peer
            case Aquainted
            case Distant
        }
        
        enum CounterWeight: String, Codable {
            case Close
            case Distant
        }
    }
    
    struct Geometry: Codable {
        let type: String
        let coordinates: [Double]
    }
}

struct APIDrop: Codable {
    let id: String
    let canvas_location: [Double]
    let image_id: String?
    let portal_url: String?
    let api_geo_node: APIGeoNode?
}

struct APIServerMessage: Codable {
    let message: String
}

enum APIMediaExtension: String, Codable {
    case png
    case jpg
    case mp4
    case mov
}




//MARK: Write
struct APICreateHost: Codable {
    let id: String
    let name: String
    let description: String
    let default_location: [String]
    let live_location: [String]?
    let portrait_id: String
    let supplement_id: String
}

struct APIUpdateHost: Codable {
    let description: String?
    let live_location: [String]?
    let portrait_id: String?
    let supplement_id: String?
}

struct APIUpdateNetwork: Codable {
    let weight: APIGeoNode.Properties.Weight
}

struct APIPullTriggers: Codable {
    let trigger_ids: [String]
}

struct APICreateDrop: Codable {
    let id: String
    let canvas_location: [String]
    let image_id: String?
    let portal_url: String?
}

